File: examples/java/kora-java-config-hocon/src/main/java/io/koraframework/example/config/hocon/FooConfig.java
package io.koraframework.example.config.hocon;

import org.jspecify.annotations.Nullable;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.time.*;
import java.util.*;
import java.util.regex.Pattern;
import io.koraframework.config.common.annotation.ConfigSource;
import io.koraframework.config.common.annotation.ConfigMapper;
import io.koraframework.common.annotation.Mapping;

@ConfigSource("foo")
public interface FooConfig {

    enum EnumValue {
        ANY,
        SOME
    }

    String valueEnvRequired();

    @Nullable
    String valueEnvOptional();

    String valueEnvDefault();

    String valueRef();

    String valueString();

    UUID valueUuid();

    Pattern valuePattern();

    EnumValue valueEnum();

    LocalDate valueLocalDate();

    LocalTime valueLocalTime();

    LocalDateTime valueLocalDateTime();

    OffsetTime valueOffsetTime();

    OffsetDateTime valueOffsetDateTime();

    Period valuePeriodAsInt();

    Period valuePeriodAsString();

    Duration valueDuration();

    int valueInt();

    long valueLong();

    BigInteger valueBigInt();

    double valueDouble();

    BigDecimal valueBigDecimal();

    @Mapping(TokenConfigValueMapper.class)
    Token apiToken();

    String relaxedKey();

    boolean valueBoolean();

    List<String> valueListAsString();

    List<String> valueListAsArray();

    Set<String> valueSetAsString();

    Set<String> valueSetAsArray();

    Map<String, String> valueMap();

    Properties valueProperties();

    @ConfigMapper
    interface BarConfig {

        String someBarString();

        BazConfig baz();

        @ConfigMapper
        interface BazConfig {

            String someBazString();
        }
    }

    BarConfig bar();

    List<BarConfig> bars();
}

File: examples/kotlin/kora-kotlin-config-yaml/src/main/kotlin/io/koraframework/kotlin/example/config/hocon/FooConfig.kt
package io.koraframework.kotlin.example.config.hocon

import io.koraframework.config.common.annotation.ConfigSource
import io.koraframework.config.common.annotation.ConfigMapper
import java.math.BigDecimal
import java.math.BigInteger
import java.time.*
import java.util.UUID
import java.util.Properties
import java.util.regex.Pattern

@ConfigSource("foo")
interface FooConfig {
    enum class EnumValue {
        ANY,
        SOME
    }

    fun valueEnvRequired(): String

    fun valueEnvOptional(): String?

    fun valueEnvDefault(): String
    fun valueRef(): String
    fun valueString(): String
    fun valueUuid(): UUID
    fun valuePattern(): Pattern
    fun valueEnum(): EnumValue
    fun valueLocalDate(): LocalDate
    fun valueLocalTime(): LocalTime
    fun valueLocalDateTime(): LocalDateTime
    fun valueOffsetTime(): OffsetTime
    fun valueOffsetDateTime(): OffsetDateTime
    fun valuePeriodAsInt(): Period
    fun valuePeriodAsString(): Period
    fun valueDuration(): Duration
    fun valueInt(): Int
    fun valueLong(): Long
    fun valueBigInt(): BigInteger
    fun valueDouble(): Double
    fun valueBigDecimal(): BigDecimal
    fun valueBoolean(): Boolean
    fun valueListAsString(): List<String>
    fun valueListAsArray(): List<String>
    fun valueSetAsString(): Set<String>
    fun valueSetAsArray(): Set<String>
    fun valueMap(): Map<String, String>
    fun valueProperties(): Properties

    @ConfigMapper
    interface BarConfig {
        fun someBarString(): String
        fun baz(): BazConfig

        @ConfigMapper
        interface BazConfig {
            fun someBazString(): String
        }
    }

    fun bar(): BarConfig
    fun bars(): List<BarConfig>
}

